### ISSUE TITLE
Bzlmod: allow repos to be visible to other module extensions

### ISSUE THREAD

**Issue by tyler-french** [bazelbuild/bazel#19301]
### Description of the feature request:

In certain situations, there exists a need to have a module extension be able to have visibility of repositories that are created by other module extensions.

Let's take `go_sdk` module extension for example in `rules_go`.

I have some repo declared locally in the main module, say for example `@go_sdk_linux_amd64` is the name of the repo.

Now, I want to have the `go_sdk` module extension load this with something that is similar to the `WORKSPACE` repository rule `go_wrap_sdk`: (https://github.com/bazelbuild/rules_go/blob/master/go/private/sdk.bzl#L372-L391)

The issue here, is that if I make a tag (for example `go_sdk.wrap(name = "@go_sdk_linux_amd64")`, the build will fail because the module extension `go_deps` doesn't see the repo @go_sdk_linux_amd64
```
ERROR: An error occurred during the fetch of repository 'rules_go~override~go_sdk~go_sdk':
   Traceback (most recent call last):
        File "/home/user/.cache/bazel/_bazel_tfrench/b97476d719d716accead0f2d5b93104f/external/rules_go~override/go/private/sdk.bzl", line 366, column 26, in _go_wrap_sdk_impl
                goroot = str(ctx.path(root_file).dirname)
Error in path: Unable to load package for @go_sdk_linux_amd64//:README.md: The repository '@go_sdk_linux_amd64' could not be resolved: Repository '@go_sdk_linux_amd64' is not defined
ERROR: <builtin>: fetching go_wrap_sdk_rule rule //:rules_go~override~go_sdk~go_sdk: Traceback (most recent call last):
        File "/home/user/.cache/bazel/_bazel_tfrench/b97476d719d716accead0f2d5b93104f/external/rules_go~override/go/private/sdk.bzl", line 366, column 26, in _go_wrap_sdk_impl
                goroot = str(ctx.path(root_file).dirname)
Error in path: Unable to load package for @go_sdk_linux_amd64//:README.md: The repository '@go_sdk_linux_amd64' could not be resolved: Repository '@go_sdk_linux_amd64' is not defined
...
FAILED: Build did NOT complete successfully (77 packages loaded, 786 targets configured)
```

### What underlying problem are you trying to solve with this feature?

The ability to use a repo that is declared in the main module in a call to a module extension that is externally declared.

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 6.3.1

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
N/A non-public code
```

**Comment by shahms** [comment]
FWIW, I encountered a similar situation when attempting to modularize Kythe.  We depend on https://github.com/google/brotli from both C++ and Go.  The Go module also depends on the C++ library, but that library can't be handled directly by Go.  In order to accomplish this, I had to patch the BUILD file to use the canonical repo path to the non-modular C++ repository.  Were that repository to be a modular dependency, this gets really fiddly and flaky.

Similarly, any kind of shared dependency that needs to bridge external package managers becomes challenging without someway of making that dependency known and available to the language-specific module extension.

**Comment by fmeum** [comment]
While this doesn't help in the general case, in the Go case you can just add the Go module with patches to the BCR as a Bazel module and `go_deps` will pick it up as a Go dependency automatically.

But I fully agree we need something more flexible and generally applicable to add the repo mapping entries for patches.

**Comment by Wyverald** [comment]
> The issue here, is that if I make a tag (for example go_sdk.wrap(name = "@go_sdk_linux_amd64"), the build will fail because the module extension go_deps doesn't see the repo @go_sdk_linux_amd64

This can work as long as the `name` attribute is of type label, not string. If the MODULE.bazel in which you're writing that tag has visibility to a repo with the apparent name "go_sdk_linux_amd64", that label would be properly translated into one with a canonical repo name which the `go_sdk` extension can freely use. I recommend (re-)reading https://bazel.build/external/overview#concepts (especially the two items about repo names) to understand more fully how this works.

Similar for @shahms, as long as you make sure the tag attribute has type "label", you should be able to pass that label around (in a capability-passing fashion, almost).

**Comment by fmeum** [comment]
> Similar for @shahms, as long as you make sure the tag attribute has type "label", you should be able to pass that label around (in a capability-passing fashion, almost).

@shahms use case, which is similar to https://github.com/bazelbuild/rules_python/issues/48#issuecomment-1669055526, also requires a way to reference this label from (patched) BUILD files.

I can think of three ways to make this work:
1. The `brotli` dependency is patched into `gazelle` as a `bazel_dep` so that the repositories generated by `go_deps` can reference it under a stable, friendly name. This requires patching both `gazelle` as well as the `go_deps`-generated repo. If `brotli` were fetched by some other module extension, the required `bazel_dep` and extension usage for that repo would also need to be patched into `gazelle`. Compared to WORKSPACE, this requires patching an additional repository.
2. `go_deps` needs to be extended with some kind of a `label_keyed_string_dict` attribute that matches canonical repo names to user-provided stable identifiers and textually replaces the latter with the former in all patch files (e.g. replace `@brotli` with `@@<brotli's canonical name>`). This would make existing patches work with Bzlmod, but textually replacing labels in patch files may turn out to be brittle.
3. Instead of textually replacing apparent with canonical labels, use the `label_keyed_string_dict` to generate a `.bzl` file with a `patch_label` function that is visible to all `go_deps`-generated repos and that patches can wrap their apparent labels into to get a canonical label string. Compared to WORKSPACE, this would make patches longer as they would also need to patch in a `load`, but isn't brittle like 2.

**Comment by fmeum** [comment]
I found a fourth option that I currently like best:
4. Provide a `repo_override` tag on the module extension that accepts a `repo_name` and a list of `targets` and creates special repo named `repo_name` that contains aliases to all the `targets` under the same respective packages. Since this is an actual repo called `repo_name`, all other extension repos will see it under this name. Only downside is that the user has to manually list all the relevant targets.

**Comment by Wyverald** [comment]
So IIUC the capability is there (using label-typed attributes), just the usability is _really_ bad because the user needs to somehow translate a BUILD file patch containing single-`@` labels to a patch containing double-`@` labels. Is that correct?

Assuming that it is, I'm sympathetic to the concern and agree that we should have a better story for this use case.

But first I wanted to make sure we're aware of a specific technical challenge, which is that any "repo mapping injection" we do is subject to conflicts. Suppose `foo` and `bar` are both using gazelle to bring in `go-brotli`, and would like to inject `cc-brotli`, we could have a conflict on our hands if `foo` and `bar` present different `cc-brotli`s to gazelle. We could of course say that only the root module can perform injections, or that conflicts are the user's business, but it's something we need to acknowledge.

---

Now, going over @fmeum's suggestions:

1. This seems extremely ugly to me, and then again could only be done by the root module, *and* it can only work if you use a non-registry override on gazelle (single_version_override wouldn't work because we don't apply the patch until we fetch the repo, and the MODULE.bazel file of gazelle is obtained from the registry, not from the [patched] fetch).

2. This is indeed too brittle IMO (and suffers from the injection conflict problem).

3. Makes writing a BUILD patch _very_ unintuitive (and suffers from the injection conflict problem).

4. This is essentially a "repo alias" in its most ideal form, which ends up being a really obvious way to work around the deficiencies of the current repo mapping rules -- oh I can't map A to B? let me just create a C which is an alias of B and map A to C instead... at that point I'd rather fix it from Bazel's side. (And... you guessed it... this suffers from the injection conflict problem...)

The "reverse use_repo"/"make_visible_to" idea brought up in the other thread also suffers from the injection conflict problem, but has the advantage of being more ergonomic -- no funny label business.

---

To mitigate the injection conflict problem, I wonder if we can limit the scope of an injected repo mapping to select repos generated by the extension. Essentially this would be bringing back the `repo_mapping` faux-attribute on repo rules to all extension impl functions, but with semantics more akin to `additional_repo_mappings`. Then the user can pass labels into the extension, and the extension can selectively make those available to repos it generates (and decide what to do when two injections conflict).

But wait, isn't that https://github.com/bazelbuild/bazel/issues/17493#issuecomment-1438420949, where you said this was infeasible?? Well, not quite -- the `repo_mapping` attribute proposed in that issue changes the repo mapping of the _Bazel modules_ using the extension, not the repos generated by the extension. By contrast, `additional_repo_mappings` only affects the repo mapping of extension-generated repos, which is already only computable after extensions run, so slapping some extra entries on there isn't a huge deal.

This approach still requires the user to "expose" a repo to an extension via a label (a capability), but then the extension can decide smartly what to do with that capability. Admittedly usability still isn't great; we _could_ repurpose the "make_visible_to" idea to improve usability (`module_ctx.modules[0].exports`?), but I'd like to hear more thoughts first.

**Comment by meteorcloudy** [comment]
In the internal team discussion we had on how to pin a module extension generated repo, instead of using `--override_repository`, I had the following proposal of a `repo_override` directive in MODULE.bazel which can override the repospec and repo mapping of a given repo.

```
# Assuming I want to pin repo foo and bar in the following MODULE.bazel file

# MODULE.bazel
bazel_dep(name = "foo", version = "1.0")

ext = use_extension("//:extension.bzl", "my_extension")
ext.bar(name = "bar", ...)
use_repo(ext, "bar")

# After vendoring foo and bar, copy then out to <source dir>/third_party (or some arbitrary dir name)

# Introduce repo_override directive in MODULE.bazel and append the following content to MODULE.bazel

repo_override(
  repo_name = "foo", # The apparent repo name that's introduced by bazel_dep or use_repo
  repo_class = "local_repository",
  # Can it just be `path = "./third_party/foo",` ?
  attrs = {"path" = "./third_party/foo"},
)

repo_override(
  repo_name = "bar",
  repo_class = "local_repository",
  attrs = {"path": "./third_party/bar"},
  # If the overridden version requires additional repo mappings, users can fix them here.
  # The key is the apparent repo name seen by bar, and the value is the apparent repo name seen by the root module.
  additional_repo_mapping = {"baz": "my_barz"}
)

# repo_override will only be available in the root module like other overrides.
# Besides solving the pinning problem for vendor mode, it also solves the problem of overriding a module extension introduced repo.

```

If we only use the repo mapping functionality alone, for the go use case, it would be:
```
repo_override(
  repo_name = "go_sdk",
  # The first `go_sdk_linux_amd64` is the apparent repo name seen by `go_sdk`
  # The second `go_sdk_linux_amd64` is the **apparent repo name** seen by the root module
  additional_repo_mapping = {"go_sdk_linux_amd64": "go_sdk_linux_amd64"}
)
```

**Comment by iancha1992** [comment]
@bazel-io fork 7.1.0

**Comment by Wyverald** [comment]
This still needs some design work, so it's unlikely to happen for 7.1.0. Punting to 7.2.0 for now.

**Comment by sluongng** [comment]
I attempted to migrate https://github.com/buildbuddy-io/buildbuddy to bzlmod yesterday and I am blocked by this.

My current understanding is that on top of the existing targets/packages visibility setting, bzlmod is applying a new visibility setting for repos. Repos are classified into 4 groups:

1. The main repo: our `@buildbuddy` repo
2. Bazel module repos: the downloaded modules we declared in MODULE.bazel file
3. Module extension repos: the repositories generated from module's extension. I.e. `@gazelle` is a module we depend on. It exposes a `go_deps` extension that generates go_repository repos from `go.mod` file.
4. Repos declared in WORKSPACE file (for backward compatibility and migration support).


I observed the following problems:

## Repos in (3) cannot depend on (1)

We have custom gazelle directives that generate BUILD files in (3) with dependencies on targets in (1) like this `@buildbuddy//some:target`. Build would fail with this class of error:

```
ERROR: no such package '@@[unknown repo 'buildbuddy' requested from @@gazelle~0.35.0~go_deps~com_google_cloud_go_longrunning]//proto': The repository '@@[unknown repo 'buildbuddy' requested from @@gazelle~0.35.0~go_deps~com_google_cloud_go_longrunning]' could not be resolved: No repository visible as '@buildbuddy' from repository '@@gazelle~0.35.0~go_deps~com_google_cloud_go_longrunning'
```

I was able to apply a workaround where I moved `@buildbuddy//some:target` into a repo in (3), but it resulted in some massive refactoring on our end to update the target in different places.

@fmeum kindly let me know that I could change it to `@@//some:target` to bypass the visibility constraint. I have yet to test this, but if this is the expected usage pattern, I would appreciate some docs for it.

## Repos in (3) cannot depend on module repo in (2)

We have some go repos that depend on `@zlib//` to compile. This would error out with

```
ERROR: no such package '@@[unknown repo 'zlib' requested from @@gazelle~override~go_deps~com_github_awslabs_soci_snapshotter]//': The repository '@@[unknown repo 'zlib' requested from @@gazelle~override~go_deps~com_github_awslabs_soci_snapshotter]' could not be resolved: No repository visible as '@zlib' from repository '@@gazelle~override~go_deps~com_github_awslabs_soci_snapshotter'
```

I have yet to find a workaround for this.

## Repos in (3) cannot depend on repo declared in (4)

I have `rules_nodejs` declared the old fashion way in `WORKSPACE.bzlmod` file.
When I try to build a target depending on `@com_github_bazelbuild_bazelisk//:bazelisk`, I would get this.

```
ERROR: error loading package '@@gazelle~override~go_deps~com_github_bazelbuild_bazelisk//': Unable to find package for @@[unknown repo 'build_bazel_rules_nodejs' requested from @@gazelle~override~go_deps~com_github_bazelbuild_bazelisk]//:index.bzl: The repository '@@[unknown repo 'build_bazel_rules_nodejs' requested from @@gazelle~override~go_deps~com_github_bazelbuild_bazelisk]' could not be resolved: No repository visible as '@build_bazel_rules_nodejs' from repository '@@gazelle~override~go_deps~com_github_bazelbuild_bazelisk'.
```

I could `bazel query '@build_bazel_rules_nodejs//...'` successfully, so this is likely another visibility issue.

**Comment by sluongng** [comment]
A minor visibility issue that I reported in Slack as well

Given

```
module(name = "buildbuddy")

...

bazel_dep(name = "gazelle", version = "0.35.0", repo_name = "bazel_gazelle")
archive_override(
    module_name = "gazelle",
    integrity = "sha256-MpOL2hbmcABjA1R5Bj2dJMYO2o15/Uc5Vj9Q0zHLMgk=",
    patch_strip = 1,
    patches = [
        "@buildbuddy//buildpatches:gazelle.patch",
    ],
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/bazel-gazelle/releases/download/v0.35.0/bazel-gazelle-v0.35.0.tar.gz",
        "https://github.com/bazelbuild/bazel-gazelle/releases/download/v0.35.0/bazel-gazelle-v0.35.0.tar.gz",
    ],
)
```

I would get

```
ERROR: An error occurred during the fetch of repository 'gazelle~override':
   Traceback (most recent call last):
        File "/private/var/tmp/_bazel_sluongng/06e573a93bc2d6a9cad4ad41f00b4310/external/bazel_tools/tools/build_defs/repo/http.bzl", line 156, column 10, in _http_archive_impl
                patch(ctx, auth = auth)
        File "/private/var/tmp/_bazel_sluongng/06e573a93bc2d6a9cad4ad41f00b4310/external/bazel_tools/tools/build_defs/repo/utils.bzl", line 174, column 22, in patch
                ctx.patch(patchfile, strip)
Error in patch: Unable to load package for @@[unknown repo 'buildbuddy' requested from @@]//buildpatches:gazelle.patch: The repository '@@[unknown repo 'buildbuddy' requested from @@]' could not be resolved: No repository visible as '@buildbuddy' from main repository
```

A workaround is to apply `s/@buildbuddy/@/` or `s/@buildbuddy/@@/`. It's a bit weird that we cannot use `@<workspace-name>` to refer to our patches.

**Comment by fmeum** [comment]
The visibility of (3) into (1) and (2) (and perhaps even (4)) is something we need to make configurable in some way as part of this issue.

The particular case of (4) not being visible to (3) is arguably an issue with the Bazelisk repo: It's not a pure Go repo, but references other Bazel rulesets in its build files. As a result, it would have to provide a `MODULE.bazel` file to declare these Bazel dependencies. Then, if you add it as a `bazel_dep`, Gazelle will pick it up automatically. I would recommend filing an issue in the Bazelisk repo for this.

Your last observation about the module name not being usable in labels in the MODULE file has a more straightforward fix. I will look into it.

**Comment by fmeum** [comment]
@tyler-french's use case is best addressed by an `attr.label` attribute on a tag class. Conveniently, strings of the form `@some_repo_name` are valid labels (shorthand for `@some_repo_name//:some_repo_name`), so the user-facing bits are identical between WORKSPACE and Bzlmod. We don't need to add any new features for this.

The remaining use cases all seem to be about tags attributes referencing patch files or inline Starlark content (think `build_file_content`). Since it's impossible to meaningfully reconcile multiple values for such attribute, they should be limited to root module and isolated usages anyway. This suggests the following type of API that is simpler than a fully generic "make A visible to B" override concept:

1. When a module has an override with `patches` (or `patch_cmd`, ...) set to a non-default value, add the root module repo mapping's entries to the module's repo mapping with lower precedence than the original entries.
2. Allow extensions to set a boolean-valued `_sees_root_module_repos` attribute on any repo rule they instantiate, which in the same way adds the root module's repo mapping (or the using module in the case of an isolated extension). This bit can be persisted in the lockfile as part of the `RepoSpec` and can be set by module extensions on repos that carry something equivalent to a `patches` attribute.

Crucially, with this approach users don't have to repeat the repositories referenced by patches, which would be additional overhead compared to WORKSPACE. 

Since a module is assumed to work with its own repo mapping, adding entries of lower precedence should not break it or hide visibility issues. In case of an unintended name collision, users can always mitigate it by changing their own `use_repo` or by applying further patches.

@meteorcloudy @Wyverald what do you think?

**Comment by sluongng** [comment]
I found a workaround for the `@zlib` issue, (3) depends on (2), above.

I smuggled the build target to the main repo (1) with
```
alias(
  name = "zlib",
  actual = "@zlib",
)
```

then I applied the workaround for the "(3) depending on (1)" issue:

```diff
- @zlib
+ @@//:zlib
```

and it worked.

This build target smuggling business made me feel like an [Onion Knight](https://en.wikipedia.org/wiki/Davos_Seaworth) and tis but a silly business. These "creative usage patterns" could be a side effect of the constraint not motivating the right behavior in the longer run and lead to a more convoluted build targets layout.

**Comment by rkallal** [comment]
Our specific use case is custom rules for internal programming languages where the rules are in a on-prem bzlmode registry i.e. REPO(2)s. We then use module extensions i.e. REPO(3)s to pull directory trees from our internal component management system and all of those "components" are "new" to bazel. We end patching BUILD.bazel files and fixes over all those "new" to bazel "components". 

We are working towards a proxy approach where there is a 1:1 between our REPO(2) for every one of our  "component" REPO(3)s.  Questions:

1) Could the _sees_root_module_repos be _sees_module_repos of a list of REPO(2)s? I may be completely missing some details because our team is new to bazel. I am presuming that root can't see all REPO(2)s that other REPO(2)s bring in via bazel_dep but root itself does not bring in. 

2)  Would you consider negotiating the bazel_dep REPO versions but allow a REPO(2) to be actually populated using a module extension REPO such that the external/REPO(2)~version_number directory is actually filled from a module extension REPO?

**Comment by fmeum** [comment]
> Could the _sees_root_module_repos be _sees_module_repos of a list of REPO(2)s? I may be completely missing some details because our team is new to bazel. I am presuming that root can't see all REPO(2)s that other REPO(2)s bring in via bazel_dep but root itself does not bring in.

Every module extension repo sees all repos that the module repo containing the extension sees. If I understood your setup correctly, this should be sufficient for your use case as you control all module repos anyway. You don't seem to even need any new feature for this.

> Would you consider negotiating the bazel_dep REPO versions but allow a REPO(2) to be actually populated using a module extension REPO such that the external/REPO(2)~version_number directory is actually filled from a module extension REPO?

I don't think I fully understand your setup yet. Why do you want to populate a module from an extension repo? It sounds like you either want a private registry (you can even have one backed by a local directory) or just reference extension repos directly and specify their Bazel deps in the module declaring the extension.

**Comment by rkallal** [comment]
> > Could the _sees_root_module_repos be _sees_module_repos of a list of REPO(2)s? I may be completely missing some details because our team is new to bazel. I am presuming that root can't see all REPO(2)s that other REPO(2)s bring in via bazel_dep but root itself does not bring in.
> 
> Every module extension repo sees all repos that the module repo containing the extension sees. If I understood your setup correctly, this should be sufficient for your use case as you control all module repos anyway. You don't seem to even need any new feature for this.

We acknowledge that when each module extension generates a repository rule (sharing the same name and extension.bzl file), it results in repositories that have visibility to one another.

Today in our user's MODULE.bazel files, we have this:
```
# An on-prem bzlmod managed repository for the verilog language.  
bazel_dep(name = "verilog_rules", version = "0.0.1")  
#An on-prem bzlmod managed repository for our internal source management system.
bazel_dep(name = "cbwa", version = "0.0.1) # 
cbwa = use_extension("@cbwa//:extensions.bzl", "cbwa")

# An on-prem component named "foo" which is managed by an internal source management system.  
cbwa.component(name = "foo") # This a REPO(2) type  
use_repo(cbwa, "foo") # This will create:_bazel_build_directory_/external/cbwa~0.0.1~cbwa~foo
```

We want our users to be able to do this without having to think about where that component _foo_ comes from.
```
# An on-prem bzlmod managed repository for the verilog language.  
bazel_dep(name = "verilog_rules", version = "0.0.1") 
# An on-prem bzlmod managed module which is acting as a proxy for the code above. 
bazel_dep(name = "foo", version = "0.0.1") 
```

We currently achieve this functionality by having both a `_bazel_build_directory_/external/foo~0.0.1` and a `_bazel_build_directory_/external/cbwa~0.0.1~cbwa~foo`. The build rules are placed in `foo~0.0.1` (i.e., the bzlmod registry-managed version of foo), while the source files reside in `_bazel_build_directory_/external/cbwa~0.0.1~cbwa~foo`. We then patch `_bazel_build_directory_/external/cbwa~0.0.1~cbwa~foo` to export all its source files. Although it works, the approach is considered messy by the teams owning the task.

> > Would you consider negotiating the bazel_dep REPO versions but allow a REPO(2) to be actually populated using a module extension REPO such that the external/REPO(2)~version_number directory is actually filled from a module extension REPO?
> 
> I don't think I fully understand your setup yet. Why do you want to populate a module from an extension repo? It sounds like you either want a private registry (you can even have one backed by a local directory) or just reference extension repos directly and specify their Bazel deps in the module declaring the extension.

The bzlmod system currently supports various suppliers such as HTTP servers and local directories, enabling HTTP or local directories to provide source files for populating the __bazel_build_director_/external/_repositoryName_ directories. Our proposal is to extend the suppliers by allowing one of these source file suppliers to be a Bazel repository rule itself. Consider a new option in MODULE.bazel files like the following:
```
module(name = "foo")
bazel_dep(name = "cbwa", version = "0.0.1) # 
cbwa = use_extension("@cbwa//:extensions.bzl", "cbwa")
cbwa.component(name = "foo") # This a REPO(2) type  

# This would tell bzlmod that the actual source is really a REPO(2) type. 
repo_source(cbwa,"foo")
```

**Comment by fmeum** [comment]
@rkallal This sounds like a request to add support for arbitrary repo rules as `source.json` type, which isn't necessarily related to repo visibility. Could you create a separate issue for this?

**Comment by bfsoyc** [comment]
I was blocked by the similar issue.
Here is my scenario: I try to build protobuf with cpp impl backend (with flag --define=use_fast_cpp_protos=true)

The protobuf require python toolchain for compiling. With legacy WORKSPACE:
```star lark
# .bazelrc
build --define=use_fast_cpp_protos=true
```

```star lark
# WORKSPACE
load("@rules_python//python:repositories.bzl", "python_register_toolchains")

# Register toolchain for hermetic build
python_register_toolchains(
    name = "python_3_9",
    python_version = "3.9",
)

# Use bind to override the //external: python_headers used in protobuf repo
bind(
    name = "python_headers",
    actual = "@python_3_9//:python_headers",
)

http_archive(
  name = "com_google_protobuf",
  # XXXX
)
```

Now the newest protobuf in BCR adapt for bazlmod by replacing use of `//external: python_headers` to `//third_party: python_headers`, is there anyway to override it?
```star lark
# MODULE.bazel
python = use_extension("@rules_python//python/extensions:python.bzl", "python")

python.toolchain(
    configure_coverage_tool = True,
    is_default = True,
    python_version = "3.9",
)
use_repo(python, "python_3_9")

bazel_dep(name = "protobuf", version = "23.1", repo_name = "com_google_protobuf")
# !!!!!!!!!!! How to let the protobuf module adapt my registers python toolchain? !!!!!!!!!!!!
```
Any help will be appreciate, thanks.

**Comment by bfsoyc** [comment]
> I was blocked by the similar issue. Here is my scenario: I try to build protobuf with cpp impl backend (with flag --define=use_fast_cpp_protos=true)
> 
> The protobuf require python toolchain for compiling. With legacy WORKSPACE:
> 
> ```star
> # .bazelrc
> build --define=use_fast_cpp_protos=true
> ```
> 
>  ```star
> # WORKSPACE
> load("@rules_python//python:repositories.bzl", "python_register_toolchains")
> 
> # Register toolchain for hermetic build
> python_register_toolchains(
>     name = "python_3_9",
>     python_version = "3.9",
> )
> 
> # Use bind to override the //external: python_headers used in protobuf repo
> bind(
>     name = "python_headers",
>     actual = "@python_3_9//:python_headers",
> )
> 
> http_archive(
>   name = "com_google_protobuf",
>   # XXXX
> )
> ```
> 
> Now the newest protobuf in BCR adapt for bazlmod by replacing use of `//external: python_headers` to `//third_party: python_headers`, is there anyway to override it?
> 
> ```star
> # MODULE.bazel
> python = use_extension("@rules_python//python/extensions:python.bzl", "python")
> 
> python.toolchain(
>     configure_coverage_tool = True,
>     is_default = True,
>     python_version = "3.9",
> )
> use_repo(python, "python_3_9")
> 
> bazel_dep(name = "protobuf", version = "23.1", repo_name = "com_google_protobuf")
> # !!!!!!!!!!! How to let the protobuf module adapt my registers python toolchain? !!!!!!!!!!!!
> ```
> 
> Any help will be appreciate, thanks.

I just found out my build ERROR is caused by https://github.com/pybind/pybind11_bazel/issues/57#issuecomment-1895839578
"Bazel 7 broke pybind11_bazel on macOS"
Add `--linkopt="-undefined dynamic_lookup"` is a workaround.
Never mind my question.  Protobuf repo defines itself well. 
Please educate me if you have more to say. 😭

**Comment by alexeagle** [comment]
This seems like a fatal flaw for rules_go users, in the sense that bzlmod requires me to go visit the BUILD files of any Go package I depend on, and alter them to conform to the strict repo visibility semantics. @Wyverald is it still on-track for 7.2?

**Comment by Wyverald** [comment]
> is it still on-track for 7.2?

Yes. Current plan is to do https://github.com/bazelbuild/bazel/issues/19301#issuecomment-1708039047. Does that address your concern?

**Comment by alexeagle** [comment]
It looks to me like that proposal still requires that users of a go dependency will have to include workarounds in their MODULE.bazel file? Or will rules_go be able to use that API internally such that every repository it generates works correctly? Probably a question for @fmeum

**Comment by fmeum** [comment]
While writing a module extension that instantiates repos with user-supplied BUILD files, I used the following replacement for `repo_override`: I added an `attr.label_list()` called `bazel_dep` that accepts labels of the form `@rules_java`. Conveniently, when such a label goes through repo mapping, it will be of the form `@@rules_java~1.2.3//:rules_java`, i.e., you can get both the canonical and the apparent name from it. This information can then be used to rewrite labels in BUILD files (for which textually replacing `"@rules_java//` and `"@rules_java"` and is probably okay).

`repo_override` or alternatives would still be much more reliable and powerful.

**Comment by sluongng** [comment]
FYI in the upcoming Gazelle release, there is a new workaround for this issue https://github.com/bazelbuild/bazel-gazelle/commit/7d10bf7befce477f518056cfc502939000bea3fa

Using the new mode, for a go_repository, gazelle will wipe all existing BUILD files and generate them new from scratch. This ensures that only MODULEs that are relevant to go builds are included in each BUILD file.

The fix is not ideal, but should address most simple cases in the Go ecosystem. I hope Bazel 7.3 release will provide a solution to this madness.

**Comment by fmeum** [comment]
If we decouple "overriding a repo definition" from "modifying a repo's repo mapping", we may be able to realize a cleaner API for the latter. This is a concept with different pros and cons:

```starlark
use_repo(some_ext, my_repo_a = "repo_a")
use_repo(other_ext, my_repo_b = “repo_b”, my_repo_c = "repo_c")

repo_mapping_override(
    "my_repo_a", # required positional argument
    “my_repo_b”, # extra positionals
    io_bazel_repo_c = "my_repo_c", # extra keyword arguments
)

# vs. 
repo_override(
    repo_name = "my_repo_a",
    additional_repo_mapping = {
        "my_repo_b": "my_repo_b",
        “io_bazel_repo_c”: “my_repo_c”,
    },
)
```

Pros I see compared to the `repo_override` approach:
* Overall less verbose.
* Makes the "what is seen as what" part less symmetric in a way that corresponds to `use_repo`, thus making it easier to remember what goes where.
* In the common use case where the user needs to make a repo visible that they are referencing in a patch, the extra positionals form can be used and removes the need to duplicate the repo name. This is particularly useful as typos in the `additional_repo_mapping` keys are harder to spot (although we could add "did you mean" suggestions to the non-visible form of a `RepositoryName` to help with this).

Cons:
* The first positional argument has a different meaning than the other ones, but the same type. This could possibly be mitigated somewhat by better naming (e.g. `make_visible_to`), but that could be hard if we want to keep the `_override` suffix.

Generally speaking, I find the `repo_override` API easier to grok if you have very little idea what it's about due to it being more verbose and `repo_mapping_override`'s API more pleasant if you work with it or wonder what it's doing *exactly*.

**Comment by Wyverald** [comment]
> If we decouple "overriding a repo definition" from "modifying a repo's repo mapping", we may be able to realize a cleaner API for the latter.

Does this mean we'd need to introduce two new directives? One for repo definitions, one for repo mappings? (Since as we discussed elsewhere, I don't think we should reuse `archive_override` & co for overriding _repo_ definitions.)

But now that I think about it, we could actually do this decoupling and make "modifying a repo's repo mapping" work on a per-extension level. The "overriding a repo definition" part simply becomes "define a repo using any mechanism, be it `use_repo` or `use_repo_rule`" (or maybe even `bazel_dep`, if we're feeling wild).

This even works around the "con" you named, since the first argument would just be the extension proxy object.

So, for Yun's earlier example:

```bzl
# This earlier example:
ext = use_extension("//:extension.bzl", "my_extension")
ext.bar(name = "bar", ...)
use_repo(ext, "bar")
repo_override(
  repo_name = "bar",
  repo_class = "local_repository",
  attrs = {"path": "./third_party/bar"},
  # If the overridden version requires additional repo mappings, users can fix them here.
  # The key is the apparent repo name seen by bar, and the value is the apparent repo name seen by the root module.
  additional_repo_mapping = {"baz": "my_barz"}
)

########################
# ... would become:
local_repository = use_repo_rule("@bazel_tools//:blah.bzl", "local_repository")
local_repository(name="my_own_bar", path="./third_party/bar")

ext = use_extension("//:extension.bzl", "my_extension")
ext.bar(name = "bar", ...)
repo_override(ext, bar="my_own_bar", baz="my_barz")
```

One extra good thing: You don't need to `use_repo(ext, "bar")` just to override it, if you're not actually using `bar` at all.

One debatable thing: Is applying a repo mapping override across all repos in an extension actually desirable? In other words, can we claim "it's never a good idea to override repo mappings for _some_ but not _all_ repos generated by an extension"? I think the answer is yes, but I'm happy to be convinced otherwise.

**Comment by meteorcloudy** [comment]
@Wyverald I like this idea. 

> repo_override(ext, bar="my_own_bar", baz="my_barz")

IIUC, this basically means the old repo is completely replaced by the new one and won't be visible to anyone, right? If so, it's essentially the same the old repo_override, but with a much cleaner API. I do tend to agree it's a much more common case to override a repo for everyone than for a specific repo.

**Comment by fmeum** [comment]
Syntax-wise `repo_override(ext, ...)` is just great - it's concise and reuses the same concepts that users know from `use_repo` while distinguishing different argument roles by types.

To clarify your idea, is it that `repo_override(ext, bar = "my_own_bar")` means "the repo mapping of repos generated by the extension `ext` maps the apparent name `bar` to the canonical repo name associated to the apparent name `my_own_bar` in the root module"?

Assuming that is the case, I have some concerns:
1. If we allow this mechanism to not only add repo mapping entries, but replace them (which is needed if we want this to provide the functionality of replacing repos), users need to keep in mind that they are replacing *some references to*, not *the definition of* the repo. If `ext` lives in `rules_foo`, generates `bar` and then has the override applied, all other repos generated by `ext` would use `my_own_bar` instead, but if `rules_foo` itself contains `use_repo(ext, "bar")`, it would use the original `bar`. Wouldn't this make it impossible to e.g. replace non-module dependencies brought in via a "WORKSPACE macro turned module extension", which are usually only referred to via `use_repo`s in the hosting module?
I wonder if we could resolve this by also mapping all `use_repo(ext, "bar")` to `my_own_bar`, thus realizing what @meteorcloudy expressed with "this basically means the old repo is completely replaced by the new one and won't be visible to anyone".
2. This technique can't add to the repo mapping of a module. Note that the seemingly simple suggestion "just patch its MODULE.bazel file" isn't actually that simple since patches on the module file supplied via `single_version_override` are *not* effective (which reminds me that we should do something about this gotcha, e.g. have a warning shown). Users would have to reach for a non-registry override and thus replace the entire module at once.
If we don't find another solution, we could always introduce an `additional_repo_mapping` attribute on `single_version_override`, just like in the original `repo_override` idea.

> In other words, can we claim "it's never a good idea to override repo mappings for some but not all repos generated by an extension"?

I could see cases where someone would want to override the mapping for a single repo (e.g. have one particular Go dependency link against a particular OpenSSL version, without affecting any other Go deps), but I don't think that's common enough to make the API for it more complicated. There is also always a workaround, assuming the module extension allows applying patches: You can first patch the references in the one extension repo to refer to a unique, new repo name and then inject a repo mapping entry for it. This seems acceptable for the rare cases in which one would need this and module extensions usually support patches.

**Comment by fmeum** [comment]
There is another type of reference which a `repo_override` based on modifying extension repo mappings wouldn't modify: If a repo rule wants to access a repo created by the same extension, it currently needs to go through a canonical name [as there is no API for this](https://github.com/bazelbuild/bazel/issues/19055), for example by trimming its own extension-provided `name` from the end and appending the name of the other repo. Such labels would still reference the "overridden" repo.

**Comment by Wyverald** [comment]
OK, so it looks like the biggest "hole" in this new idea is that the overridden repo is still accessible through its original canonical name.

This hole is rather hard to plug, since we'd need some sort of repo-level alias (https://github.com/bazelbuild/bazel/issues/19927) for this to work. (The idea being that the overridden repo's canonical name aliases to the overriding repo's canonical name.) While this is easy to do on a surface level (just make a symlink), the proper fix for this is nigh impossible (involves making labels compare equal while cognizant of repo-level aliases).

We could work around this issue in the "`use_repo` from `rules_foo`" case by making `use_repo` "smart" (as in, have it be aware of `override_repo` directives). This will probably be good for 90% of cases. For the remaining 10% where people splice repo names, we could fix #19055 and call it a day.

---

> This technique can't add to the repo mapping of a module.

While that's true, I'm still kind of sympathetic to the "patching MODULE.bazel with `single_version_override`" use case. Assuming we ever support that, it would be a "feature not a bug" that we have a single way to do this.

**Comment by fmeum** [comment]
Does that mean we have this plan?

1. Resolve #19055 and document that canonical repo names should never be constructed by hand. If rulesets don't adopt it right away, it's not the end of the world as it only leads to issues when users `repo_override` repos that have canonical labels constructed for them by other repos in the extension.
2. Implement `repo_override` including smart `use_repo`.
3. Make `single_version_override` able to patch module files, either by fetching patched modules eagerly or by magically patching only the module file.

This looks good to me (and doable in time for 7.4.0/8.0.0).

**Comment by Wyverald** [comment]
SGTM.

One last bikeshed for the road: I'd prefer the name `override_repo` over `repo_override`. `override_repo` has a nice symmetry with `use_repo`, and `repo_override` is too similar to `archive_override` & co.

**Comment by fmeum** [comment]
Some edge cases that we should keep in mind when implementing this:

```starlark
ext = use_repo("@other_module//:ext.bzl", "ext")
use_repo(ext, "foo")
# Overriding one repo in an extension with another repo from that extension.
override_repo(ext, bar = "foo")
```

```starlark
ext = use_repo("//:ext.bzl", "ext")
use_repo(ext, "foo")
ext2 = use_repo("//:ext2.bzl", "ext2")
use_repo(ext2, "bar")
# Overriding a repo in an extension defined by the main repo.
override_repo(ext, foo = "bar")
```

```starlark
ext = use_repo("@other_module//:ext.bzl", "ext")
use_repo(ext, "foo")
ext2 = use_repo("@other_module//:ext2.bzl", "ext2")
use_repo(ext2, "bar")
ext3 = use_repo("@other_module//:ext3.bzl", "ext3")
use_repo(ext3, "baz")
# Overriding a repo with an overridden repo.
override_repo(ext2, bar = "foo")
override_repo(ext3, baz = "bar")
```

```starlark
ext = use_repo("@other_module//:ext.bzl", "ext")
use_repo(ext, "foo")
ext2 = use_repo("@other_module//:ext2.bzl", "ext2")
use_repo(ext2, "bar")
ext3 = use_repo("@other_module//:ext3.bzl", "ext3")
use_repo(ext3, "baz")
# Overriding a repo with an overridden repo in a way that forms a cycle.
override_repo(ext2, bar = "foo")
override_repo(ext3, baz = "bar")
override_repo(ext, foo = "baz")
```

`override_repo` can also be used to create situations in which different apparent names map to the same canonical name in an extension's repo mapping, which is not something that could happen before. We could make this an error or just live with it.

**Comment by fmeum** [comment]
There are some aspects of `override_repo` and `inject_repo` that I'm not happy with yet. Maybe we can improve their API in time for the release. Both `override_repo` and `inject_repo` can result in multiple apparent names resolving to the same repo. This is not the case with Bzlmod if you don't use them and can be confusing.

1. For `inject_repo`, we should probably disallow `use_repo`ing the inject repo altogether:
```starlark
local_repository(
    name = "zlib",
    ...
)

foo_deps = use_extension(..., "foo_deps")
inject_repo(foo_deps, com_github_zlib = "zlib")
# Disallow this:
use_repo(foo_deps, "com_github_zlib")
```

In this example, instead of having `zlib` and `com_github_zlib` in scope as aliases of each other, it would be cleaner to just refer to `zlib` throughout.

2. For `override_repo`, the situation is more complicated (I'm using `googleapis` to stand in for some more complex module that contains targets for multiple languages):

```starlark
local_repository(
    name = "googleapis",
    ...
)

bar_deps = use_extension(..., "bar_deps")
override_repo(bar_deps, "googleapis")
use_repo(bar_deps, bar_googleapis = "googleapis")

foo_deps = use_extension(..., "foo_deps")
override_repo(foo_deps, com_github_google_googleapis = "googleapis")
use_repo(foo_deps, "com_github_googleapis")
```

Assuming that `googleapis` is considered a direct dep by both `bar_deps` and `foo_deps`, `bazel mod tidy` would enforce an import of `googleapis` from `bar_deps`, which leads to an error that the user can resolve manually by using the keyword form.

Since `bar_googleapis` and `com_github_googleapis` are just aliases for `googleapis`, it would be much cleaner if we only had `googleapis` in scope and users would consistently reference that (the overriding repo). We *could* teach `bazel mod tidy` to ignore or even remove overridden repos, resulting in this module file:

```starlark
bazel_dep(
    name = "googleapis",
    ...
)

bar_deps = use_extension(..., "bar_deps")
override_repo(bar_deps, "googleapis")

foo_deps = use_extension(..., "foo_deps")
override_repo(foo_deps, com_github_google_googleapis = "googleapis")
```

This looks much nicer, but it's flawed in a subtle way: When the module is not the root module, the overrides won't be applied and all references to `@googleapis` will fail, even if the override isn't essential to any functionality provided by the module to its consumers. Note that this wasn't a problem with `inject_repo` since the injected repo can't be referenced in any way in a setting where the module isn't the root module.

I first thought that we could prohibit `use_repo` for overridden repos just as for injected repos, but additionally turn `override_repo(bar_deps, A = "B")` into `use_repo(bar_deps, B = "A")` in a non-root module context. But that fails in the situation above as `googleapis` overrides two different repos.

I don't see a better solution than having `bazel mod tidy` neither add nor remove `use_repo` arguments for overridden repos. Users can then either use the overriding repo throughout for a cleaner repo mapping or, if needed in more complex cases, can `use_repo` to get a name that can also be referenced in a non-root module context.

Curious to hear your thoughts about this!
